### PR TITLE
add `exec::trampoline_scheduler` to bound stack usage

### PR DESCRIPTION
**Pull request by ericniebler**
Used like:

```c++
  exec::trampoline_scheduler sched;
  auto recurse_deeply = retry(exec::on(sched, fails_alot{}));
  sync_wait(std::move(recurse_deeply));
```

### CODE REVIEW

File: include/exec/trampoline_scheduler.hpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            __self.__execute();
+            __state.__drain();
+          } else if (__current_state->__recursion_depth_ < __self.__max_recursion_depth_) {
+            ++__current_state->__recursion_depth_;
+            __self.__execute();
+          } else {
+            // Exceeded recursion limit.
+            __self.__next_ = std::exchange(
+              __current_state->__head_, static_cast<__operation_base*>(&__self));
+          }
```
Comment (ericniebler) [inline include/exec/trampoline_scheduler.hpp:85]: I admit I have shamelessly copied this implementation from libunifex (and left the Facebook copyright!). Now that I'm looking at this `start` implementation, I have Questions.

If we call `start` on one of these operation states and we've hit the recursion limit, we stuff the operation into the head of a thread-local list. The next time a trampoline operation is started on this thread, we execute all the operations in the list. Fine.

But what if we _don't_ start another trampoline operation on this thread? What becomes of the started operations sitting in the list? Program correctness could demand that this operations complete. If they don't, they can hang the program.

Am I confused?

Comment (ericniebler) [reply include/exec/trampoline_scheduler.hpp:85]: attn @kirkshoop @lewissbaker

Comment (ericniebler) [reply include/exec/trampoline_scheduler.hpp:85]: Maybe when the recursion limit is hit, we should post the work to the delegatee scheduler? And hard-error if there isn't one?

Comment (lewissbaker) [reply include/exec/trampoline_scheduler.hpp:85]: When the recursion limit is reached it posts work back to the first call to start (the first if-branch where it calls drain()).

Each time execution returns to drain() loop it resets the recursion count back to 1 before executing the next item in the queue.

**Pull request merged**